src/Microsoft.ML.Data/DataLoadSave/TransformWrapper.cs

```diff
@@ -0,0 +1,152 @@
+        {
+            _host.CheckValue(inputSchema, nameof(inputSchema));
+
+            var dv = new EmptyDataView(_host, inputSchema);
+            var output = ApplyTransformUtils.ApplyAllTransformsToData(_host, _xf, dv);
+            return output.Schema;
+        }
+
+        public void Save(ModelSaveContext ctx)
+        {
+            ctx.CheckAtModel();
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            var dataPipe = _xf;
+            var transforms = new List<IDataTransform>();
+            while (dataPipe is IDataTransform xf)
+            {
+                // REVIEW: a malicious user could construct a loop in the Source chain, that would
+                // cause this method to iterate forever (and throw something when the list overflows). There's
+                // no way to insulate from ALL malicious behavior.
+                transforms.Add(xf);
+                dataPipe = xf.Source;
+                Contracts.AssertValue(dataPipe);
+            }
+            transforms.Reverse();
+
+            ctx.SaveSubModel("Loader", c => BinaryLoader.SaveInstance(_host, c, dataPipe.Schema));
+
+            ctx.Writer.Write(transforms.Count);
+            for (int i = 0; i < transforms.Count; i++)
+            {
+                var dirName = string.Format(TransformDirTemplate, i);
+                ctx.SaveModel(transforms[i], dirName);
+            }
+        }
+
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "XF  WRPR",
+                verWrittenCur: 0x00010001, // Initial
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoaderSignature);
+        }
+
+        // Factory for SignatureLoadModel.
+        public TransformWrapper(IHostEnvironment env, ModelLoadContext ctx)
+        {
+            Contracts.CheckValue(env, nameof(env));
+            _host = env.Register(nameof(TransformWrapper));
+            _host.CheckValue(ctx, nameof(ctx));
+
+            ctx.CheckAtModel(GetVersionInfo());
+            int n = ctx.Reader.ReadInt32();
+            _host.CheckDecode(n >= 0);
+
+            ctx.LoadModel<IDataLoader, SignatureLoadDataLoader>(env, out var loader, "Loader", new MultiFileSource(null));
+
+            IDataView data = loader;
+            for (int i = 0; i < n; i++)
+            {
+                var dirName = string.Format(TransformDirTemplate, i);
+                ctx.LoadModel<IDataTransform, SignatureLoadDataTransform>(env, out var xf, dirName, data);
+                data = xf;
+            }
+
+            _xf = data;
+        }
+
+        public IDataView Transform(IDataView input) => ApplyTransformUtils.ApplyAllTransformsToData(_host, _xf, input);
+    }
+
+    /// <summary>
+    /// Estimator for trained wrapped transformers.
+    /// </summary>
+    internal abstract class TrainedWrapperEstimatorBase : IEstimator<TransformWrapper>
```

Comment:
>internal [](start = 4, length = 8)

Is it `internal` by design? I need to use it to convert Ngram transform.

Reply:
no, let's make it public for now

---
In reply to: [218629635](https://github.com/dotnet/machinelearning/pull/931#discussion_r218629635) [](ancestors = 218629635)